src/execution/compiler/operator/delete_translator.cpp

```diff
@@ -63,35 +66,33 @@ void DeleteTranslator::FinishPipelineWork(const Pipeline &pipeline, FunctionBuil
 void DeleteTranslator::DeclareDeleter(FunctionBuilder *builder) const {
   // var col_oids : [0]uint32
   SetOids(builder);
-  // var deleter : StorageInterface
-  auto *storage_interface_type = GetCodeGen()->BuiltinType(ast::BuiltinType::Kind::StorageInterface);
-  builder->Append(GetCodeGen()->DeclareVarNoInit(deleter_, storage_interface_type));
-  // @storageInterfaceInit(&deleter, execCtx, table_oid, col_oids, true)
+  // @storageInterfaceInit(&pipelineState.storageInterface, execCtx, table_oid, col_oids, true)
   const auto &op = GetPlanAs<planner::DeletePlanNode>();
-  ast::Expr *deleter_setup = GetCodeGen()->StorageInterfaceInit(deleter_, GetExecutionContext(),
+  ast::Expr *deleter_setup = GetCodeGen()->StorageInterfaceInit(si_deleter_.GetPtr(GetCodeGen()), GetExecutionContext(),
                                                                 op.GetTableOid().UnderlyingValue(), col_oids_, true);
   builder->Append(GetCodeGen()->MakeStmt(deleter_setup));
 }
 
 void DeleteTranslator::GenDeleterFree(FunctionBuilder *builder) const {
-  // @storageInterfaceFree(&deleter)
+  // @storageInterfaceFree(&pipelineState.storageInterface)
   ast::Expr *deleter_free =
-      GetCodeGen()->CallBuiltin(ast::Builtin::StorageInterfaceFree, {GetCodeGen()->AddressOf(deleter_)});
+      GetCodeGen()->CallBuiltin(ast::Builtin::StorageInterfaceFree, {si_deleter_.GetPtr(GetCodeGen())});
   builder->Append(GetCodeGen()->MakeStmt(deleter_free));
 }
 
 void DeleteTranslator::GenTableDelete(FunctionBuilder *builder) const {
-  // if (!@tableDelete(&deleter, &slot)) { Abort(); }
+  // if (!@tableDelete(&pipelineState.storageInterface, &slot)) { Abort(); }
   const auto &op = GetPlanAs<planner::DeletePlanNode>();
   const auto &child = GetCompilationContext()->LookupTranslator(*op.GetChild(0));
   NOISEPAGE_ASSERT(child != nullptr, "delete should have a child");
   const auto &delete_slot = child->GetSlotAddress();
-  std::vector<ast::Expr *> delete_args{GetCodeGen()->AddressOf(deleter_), delete_slot};
+  std::vector<ast::Expr *> delete_args{si_deleter_.GetPtr(GetCodeGen()), delete_slot};
   auto *delete_call = GetCodeGen()->CallBuiltin(ast::Builtin::TableDelete, delete_args);
   auto *delete_failed = GetCodeGen()->UnaryOp(parsing::Token::Type::BANG, delete_call);
   If check(builder, delete_failed);
   {
     // The delete was not successful; abort the transaction.
+    GenDeleterFree(builder);
```

Comment:
Wouldn't this result in a double free now that we are freeing this when we tear down the pipeline state?

Reply:
absolutely, that was outdated code, thanks